src/Microsoft.OData.Edm/Validation/ValidationRules.cs

```diff
@@ -1611,7 +1611,7 @@ public static class ValidationRules
                               var principalType = principalProperties.ElementAtOrDefault(i).Type.Definition;

```

Comment:
We'er calling `ElementAtOrDefault(i)` inside of a loop. Unless the collection is backed by an `IList`, `ElementAtOrDefault(i)` will traverse the collection up to `i` for each iteration of the loop. So, this loop is essentially `O(n^2)`.

I see that we call `Count()` on these collections earlier in the code as well. These could also be performing linear scans unless the collections are backed by arrays.

This method appears to be really inefficient, performing many redundant scans. 

It may be possible to refactor this entire method to perform a single scan the steps through both collections per iteration. This might be possible if we use enumerator.MoveNext() API directly.

we could also consider performing `ToList()` on both collections at the beginning and using `.Count` (property not method) and indexing against the lists directly. Though this allocates more memory than using enumerator directly.

Reply:
I understand that these issues have nothing to do with this PR. We may consider them separate concerns for a different PR unless you're open to fixing the perf issues in this PR.

Reply:
Thanks @habbes for this. Please see my previous comment. Can we create a new issue to handle perf concerns for these validation rules and handle them in a separate PR.